core/distributed/dd_matrix.cpp

```diff
@@ -0,0 +1,416 @@
+std::unique_ptr<DdMatrix<ValueType, LocalIndexType, GlobalIndexType>>
+DdMatrix<ValueType, LocalIndexType, GlobalIndexType>::create(
+    std::shared_ptr<const Executor> exec, mpi::communicator comm,
+    ptr_param<const LinOp> matrix_template)
+{
+    return std::unique_ptr<DdMatrix>{new DdMatrix{exec, comm, matrix_template}};
+}
+
+
+template <typename ValueType, typename LocalIndexType, typename GlobalIndexType>
+void DdMatrix<ValueType, LocalIndexType, GlobalIndexType>::convert_to(
+    DdMatrix<next_precision_base<value_type>, local_index_type,
+             global_index_type>* result) const
+{
+    GKO_ASSERT(this->get_communicator().size() ==
+               result->get_communicator().size());
+    result->local_mtx_->copy_from(this->local_mtx_);
+    result->restriction_->copy_from(this->restriction_);
+    result->prolongation_->copy_from(this->prolongation_);
+    result->set_size(this->get_size());
+}
+
+
+template <typename ValueType, typename LocalIndexType, typename GlobalIndexType>
+void DdMatrix<ValueType, LocalIndexType, GlobalIndexType>::move_to(
+    DdMatrix<next_precision_base<value_type>, local_index_type,
+             global_index_type>* result)
+{
+    GKO_ASSERT(this->get_communicator().size() ==
+               result->get_communicator().size());
+    result->local_mtx_->move_from(this->local_mtx_);
+    result->restriction_->move_from(this->restriction_);
+    result->prolongation_->move_from(this->prolongation_);
+    result->set_size(this->get_size());
+    this->set_size({});
+}
+
+
+template <typename ValueType, typename LocalIndexType, typename GlobalIndexType>
+void DdMatrix<ValueType, LocalIndexType, GlobalIndexType>::read_distributed(
+    const device_matrix_data<value_type, global_index_type>& data,
+    std::shared_ptr<const Partition<local_index_type, global_index_type>>
+        partition)
+{
+    const auto comm = this->get_communicator();
+    GKO_ASSERT_EQ(data.get_size()[0], partition->get_size());
+    GKO_ASSERT_EQ(data.get_size()[1], partition->get_size());
+    GKO_ASSERT_EQ(comm.size(), partition->get_num_parts());
+    auto exec = this->get_executor();
+    auto local_part = comm.rank();
+    auto use_host_buffer = mpi::requires_host_buffer(exec, comm);
+    auto tmp_partition = make_temporary_clone(exec, partition);
+
+    // set up LinOp sizes
+    auto global_num_rows = partition->get_size();
+    dim<2> global_dim{global_num_rows, global_num_rows};
+    this->set_size(global_dim);
+
+    size_type num_parts = comm.size();
+    array<GlobalIndexType> non_owning_row_idxs{exec};
+    array<GlobalIndexType> non_owning_col_idxs{exec};
+    device_matrix_data<value_type, global_index_type> data_copy{exec, data};
+    auto arrays = data_copy.empty_out();
+
+    exec->run(dd_matrix::make_filter_non_owning_idxs(
+        data, make_temporary_clone(exec, partition).get(),
+        make_temporary_clone(exec, partition).get(), local_part,
```

Comment:
you already have a temporary partition:
```suggestion
        data, tmp_partition.get(),
        tmp_partition.get(), local_part,
```